### PR TITLE
User ownership to nodes

### PR DESCRIPTION
**Pull request by JenySadadia**
Fixes https://github.com/kernelci/kernelci-api/issues/256

Add user ownership to nodes. 
- Add `Node.user_id` field to store user ID that created the node
- Added method to authorize user for PUT node requests
- Exclude `user_id` field from response data of node related requests i.e. GET, PUT, and POST

### CODE REVIEW

**Comment by gctucker** [comment]
Looks like this can be rebased and the skip label removed now that #265 and #266 have been merged.

**Comment by JenySadadia** [comment]
> Looks like this can be rebased and the skip label removed now that #265 and #266 have been merged.

I am working on adding `NodeOwnership` model instead of adding `Node.username` and `Node.user_groups` fields.
That way we can easily manage update operations and permissions related to ownership. That can be separate from actual node objects.

**Comment by gctucker** [comment]
What's the issue with having `user` and `group` fields directly in the Node model?

**Comment by JenySadadia** [comment]
> What's the issue with having `user` and `group` fields directly in the Node model?

Fields for ownership should be editable by authorized users only.
Any user that updates the node should not be able to update `username` field. Also, `groups` should only be updated by node owner (and maybe by admin). It is easy to have a separate model than to have these constraints in POST node.
Also, we can have different endpoints defined for this model just like the user profile model.

**Comment by gctucker** [comment]
I don't see how this makes any difference for the node model though, the API will need to check if the user sending PUT is the owner of the node or belongs to the node's group in the application layer, this can't be done at the API validation layer as the node needs to be retrieved from the database first.

**Comment by JenySadadia** [comment]
> I don't see how this makes any difference for the node model though, the API will need to check if the user sending PUT is the owner of the node or belongs to the node's group in the application layer, this can't be done at the API validation layer as the node needs to be retrieved from the database first.

As per our last discussion, I have implemented `NodeOwnership` model. The ownership object will be created in POST request. It will be used to authorize user in PUT requests.

**Comment by JenySadadia** [comment]
Also, need to add `unique_items` constraint to `NodeOwnership.user_groups`.

File: api/main.py
Comment (gctucker) [inline api/main.py:375]: BTW The linter warning here is about using `None` instead, as otherwise the same object reference keeps being used for default values

**Comment by gctucker** [comment]
The issue with this is that it's creating separate documents with references to a User and a Node, which effectively creates a typical many-to-many relationship as used in SQL databases.  This means that every time a node is being handled, we need to look up a NodeOwnership document and then maybe also a User document.  However, NoSQL databases like MongoDB aren't good for that.  Instead, the typical way of doing it is by embedding the models in the main document itself.  So we could have a NodeOwnership model but that would need to be stored as Node.ownership.

I understand why you wanted to do that, to be able to deal with POST request permissions separately for the Node fields and for NodeOwnership, but in this case we can't realistically do it this way mainly for performance reasons.

I believe the best way to handle this right now is to have a Node model with just plain fields `user` and `groups` containing the names of the owner and groups the node belongs to, but with some logic at the application level to not allow changing them when handling PUT requests (except maybe for admin users).  I'm not sure if we can make them constant fields but if we can then that would work too initially.  It would keep the implementation very simple, reduce the burden of testing and be less bug-prone, and also it will appear directly in the API generated documentation.  The limitation would be that nodes can't be assigned to another owner or groups once created, which might be entirely acceptable even in the long-term anyway as I don't think we have a known use-case where changing ownership is useful as long as all nodes are publicly readable.

**Comment by JenySadadia** [comment]
> The issue with this is that it's creating separate documents with references to a User and a Node, which effectively creates a typical many-to-many relationship as used in SQL databases. This means that every time a node is being handled, we need to look up a NodeOwnership document and then maybe also a User document. However, NoSQL databases like MongoDB aren't good for that. Instead, the typical way of doing it is by embedding the models in the main document itself. So we could have a NodeOwnership model but that would need to be stored as Node.ownership.
> 
> I understand why you wanted to do that, to be able to deal with POST request permissions separately for the Node fields and for NodeOwnership, but in this case we can't realistically do it this way mainly for performance reasons.
> 
> I believe the best way to handle this right now is to have a Node model with just plain fields `user` and `groups` containing the names of the owner and groups the node belongs to, but with some logic at the application level to not allow changing them when handling PUT requests (except maybe for admin users). I'm not sure if we can make them constant fields but if we can then that would work too initially. It would keep the implementation very simple, reduce the burden of testing and be less bug-prone, and also it will appear directly in the API generated documentation. The limitation would be that nodes can't be assigned to another owner or groups once created, which might be entirely acceptable even in the long-term anyway as I don't think we have a known use-case where changing ownership is useful as long as all nodes are publicly readable.

Hello @gctucker Thanks for the feedback.
I read some documentation from mongodb about this and it seems like the better approach here is to embed fields in the same document rather than query database at each PUT request. The ownership related information is not hard to manage in a single document and also we would need it frequently.
I will update the PR accordingly.

**Comment by gctucker** [comment]
Yes, then having a model just for that or fields directly in Node is more like an implementation detail as it'll appear pretty much in the same way in MongoDB documents.  Maybe a sub-model will create a embedded dictionary but that should be a minor thing from a performance and database design point-of-view.  Whatever makes the models and application logic better designed would work in this case.

**Comment by JenySadadia** [comment]
Updated the PR.
Re-assigned ownership details in `put_node` as of now. It will need to be tweaked once we discuss about various permissions for updating node ownership.
The basic case I can think of is, if the user is an owner then should be allowed to update `user_groups`.
Also, admin users should have permission to update that info.

**Comment by JenySadadia** [comment]
We can use `update_one` instead of `replace_one` for updating documents partially.

**Comment by gctucker** [comment]
> The basic case I can think of is, if the user is an owner then should be allowed to update `user_groups`.

I mean, from an actual use-case.  Why change the group assigned to an existing node?

**Comment by JenySadadia** [comment]
> > The basic case I can think of is, if the user is an owner then should be allowed to update `user_groups`.
> 
> I mean, from an actual use-case. Why change the group assigned to an existing node?

If the owner wants to share the node with some other group that was not added at the time of node creation.

**Comment by gctucker** [comment]
We don't have this problem as long as all nodes are publicly readable.  Changing the group would only be to allow some other group of users to update the node later on, and I'm pretty sure we can live without the possibility to do this for a long time.  In fact, if a node is in Done state then I think we've said it shouldn't be possible to update it anyway?

**Comment by JenySadadia** [comment]
> We don't have this problem as long as all nodes are publicly readable. Changing the group would only be to allow some other group of users to update the node later on, and I'm pretty sure we can live without the possibility to do this for a long time. In fact, if a node is in Done state then I think we've said it shouldn't be possible to update it anyway?

Yes, update is only possible if the node is not in "done" state. Okay. So re-assigning values would do for now.

**Comment by JenySadadia** [comment]
After I get approval on initial commits, I also need to add changes for `put_nodes`.

**Comment by JenySadadia** [comment]
> > We don't have this problem as long as all nodes are publicly readable. Changing the group would only be to allow some other group of users to update the node later on, and I'm pretty sure we can live without the possibility to do this for a long time. In fact, if a node is in Done state then I think we've said it shouldn't be possible to update it anyway?
> 
> Yes, update is only possible if the node is not in "done" state. Okay. So re-assigning values would do for now.

Found a better solution to deal with this issue. 
Pydantic provides `exclude` parameter to be used while converting the model to a dictionary. Used it to exclude node ownership fields in PUT requests.

File: api/models.py
Comment (gctucker) [inline api/models.py:256]: Oh I thought we wanted to just have one group, not a list of groups.  I guess this is fine but a bit more complicated.

File: api/models.py
Comment (gctucker) [inline api/models.py:250]: Maybe `owner` would be a bit better, but that's a detail.  We can always tweak things like this a bit later.

File: api/main.py
Comment (gctucker) [inline api/main.py]: This model doesn't exist yet, maybe it's created in a later commit.

File: api/main.py
Comment (gctucker) [inline api/main.py:383]: One thought, we could also skip this and instead accept that the owner and group may not always exist in the database.  That way, when renaming or deleting users and groups we don't have to update retroactively all the older nodes.  I think it's easy to handle as the user and group lookups are done directly in the application logic.

File: api/main.py
Comment (gctucker) [inline api/main.py]: Right, that's some diff noise with the previous commit.

File: api/main.py
Comment (gctucker) [inline api/main.py:128]: Shouldn't all nodes have an owner?

**Review by gctucker — COMMENTED** [review]
Just a few more comments, generally speaking I think this is good to go now.

File: api/models.py
Comment (JenySadadia) [reply api/models.py:250]: I agree. I will rename it.

File: api/main.py
Comment (JenySadadia) [reply api/main.py:128]: This check is due to already existing nodes do not have ownership details.
Maybe we can create a user and assign it to all the existing nodes?

File: api/main.py
Comment (JenySadadia) [reply api/main.py:383]: I don't think inserting invalid data into the database for the sake of simplicity is a good idea.

File: api/models.py
Comment (JenySadadia) [reply api/models.py]: Done.

File: api/models.py
Comment (gctucker) [reply api/models.py:257]: How is this related?  I like to think of Unix file systems, users can belong to a number of groups but files have only one group.

The current implementation seems to be working so that's fine, we can always simplify things by reducing it to a single group later on.

File: api/main.py
Comment (gctucker) [reply api/main.py:384]: It depends how we define invalid, having an non-existing object ID would be more problematic I think.  But even that we're not checking afaik, say when setting a parent node id.  Ultimately, if the user provides bad data then it's a problem on the user's side.

What's important though is that we define exactly who is in charge of the data consistency.  If we want to check that groups exist and other similar checks around node ownership then we need to do it entirely, and update all nodes when a group or user gets renamed etc.  Otherwise, we can just say it's something the user is in charge of and be fault-tolerant to return HTTP 4xx errors when doing operations on data that doesn't exist.  Having fixed user names and groups is also a way to solve this by not having the use-case where things need to be renamed and updated.

File: api/main.py
Comment (gctucker) [reply api/main.py]: Yes I think all existing nodes should be updated with `admin` user or something like that.  We can also delete old nodes as this is the staging database.

**Review by gctucker — APPROVED** [review]

**Comment by gctucker** [comment]
Tested OK on staging (no side-effects on the pipeline).

File: api/main.py
Comment (JenySadadia) [reply api/main.py]: Yes, I think deleting nodes would be easier.

**Pull request merged**
